### PR TITLE
🧾 Limit orders! 

### PR DESCRIPTION
**Pull request by jahooma**
Notion design doc: https://manifoldmarkets.notion.site/Limit-orders-a48c86cdef724d3784b43ef574d2cf3a

### CODE REVIEW

**Comment by jahooma** [comment]
First pass on bet UI (binary markets only for now).

Screen Shot 2022-06-13 at 8 11 49 PM
Screen Shot 2022-06-13 at 8 11 58 PM

**Comment by ahalekelly** [comment]
Did some API performance testing, ran POST /bet in a loop on one market

| Clients  | Throughput (bets/s) | 90th percentile latency |
| ------------- | ------------- | ------------- | 
| 1  | 1.5  | 851ms| 
| 2  | 2.0  | 1988ms| 
| 4  | 1.6  | 5016ms|

Those aren't per-client throughput numbers, that's total.

**Comment by mqp** [comment]
I have no idea how to make this work performance-wise given Firestore's constraints, but I am OK with taking an approach consistent with the rest of our software, which is roughly "experiment with features in a way that doesn't scale to more than a few users on the website at a time, and then somehow later if we determine the features are right, go rewrite them in a way that scales." As long as everyone is aware that this is what we are doing.

**Comment by ahalekelly** [comment]
Hmm does this implementation write to the contract for every limit order? Austin pointed out that if you avoid that you could bypass the bottleneck.

File: common/bet.ts [hunk 1]
```diff
@@ -25,8 +26,10 @@ export type Bet = {
   isAnte?: boolean
   isLiquidityProvision?: boolean
   isRedemption?: boolean
-
-  createdTime: number
+  // A record of each transaction that partially (or fully) fills the bet amount.
+  // I.e. A limit order could be filled by partially matching with several bets.
+  // Non-limit orders can also be filled by matching with multiple limit orders.
+  fills?: fill[]
```
Comment (mantikoros) [inline common/bet.ts:32]: The bets subcollection is going to get pretty big with all the limit order and fills being stored...

File: functions/src/place-bet.ts [hunk 2]
```diff
@@ -97,11 +129,45 @@ export const placebet = newEndpoint({}, async (req, auth) => {
...
+
+      // Deduct balance of makers.
+      // TODO: Check if users would go negative from fills and cancel those bets.
+      const spentByUser = mapValues(
+        groupBy(makers, (maker) => maker.bet.userId),
+        (makers) => sumBy(makers, (maker) => maker.amount)
+      )
+      for (const [userId, spent] of Object.entries(spentByUser)) {
+        const userDoc = firestore.collection('users').doc(userId)
+        trans.update(userDoc, { balance: FieldValue.increment(-spent) })
```
Comment (mantikoros) [inline functions/src/place-bet.ts:166]: Didn't know FieldValue.increment() was a thing.

File: common/new-bet.ts [hunk 3]
```diff
@@ -30,38 +40,185 @@ export type BetInfo = {
...
+    shares: number
+    timestamp: number
+  }[] = []
+
+  let amount = betAmount
+  let cpmmState = { pool: contract.pool, p: contract.p }
+  let totalFees = noFees
+
+  let i = 0
+  while (true) {
```
Comment (mantikoros) [inline common/new-bet.ts:164]: elegant core loop

File: common/new-bet.ts [hunk 4]
```diff
@@ -30,38 +40,185 @@ export type BetInfo = {
...
+    amount -= taker.amount
+
+    if (maker.matchedBetId === null) {
+      // Matched against pool.
+      cpmmState = maker.state
+      totalFees = addObjects(totalFees, maker.fees)
+      takers.push(taker)
+    } else {
+      // Matched against bet.
+      takers.push(taker)
```
Comment (mantikoros) [inline common/new-bet.ts:180]: there should also be a creator fee assessed on the taker which should be added to totalFees

Comment (mantikoros) [reply common/new-bet.ts:180]: I guess you can leave that for v 2.

File: web/components/limit-bets.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  contract: CPMMBinaryContract | PseudoNumericContract
+  bets: LimitBet[]
+  className?: string
+}) {
+  const { contract, bets, className } = props
+  const recentBets = sortBy(bets, (bet) => bet.createdTime).reverse()
+
+  return (
+    <Col className={clsx(className, 'gap-2 rounded bg-white')}>
+      <div className="px-6 py-3 text-xl">Limit bets</div>
```
Comment (mantikoros) [inline web/components/limit-bets.tsx:23]: I still like "limit order" more.

File: common/bet.ts [hunk 6]
```diff
@@ -35,4 +38,21 @@ export type NumericBet = Bet & {
   allBetAmounts: { [outcome: string]: number }
 }
 
+// Binary market limit order.
+export type LimitBet = Bet & {
+  orderAmount: number // Amount of limit order.
+  limitProb: number // [0, 1]. Bet to this probability.
+  isFilled: boolean // Whether all of the bet amount has been filled.
+  isCancelled: boolean // Whether to prevent any further fills.
```
Comment (mantikoros) [inline common/bet.ts:46]: "canceled" is the preferred American spelling

File: common/calculate-cpmm.ts [hunk 7]
```diff
@@ -115,6 +119,53 @@ export function calculateCpmmPurchase(
   return { shares, newPool, newP, fees }
 }
 
+export function calculateCpmmAmount(
```
Comment (mantikoros) [inline common/calculate-cpmm.ts:122]: I think there is probably a closed-form solution to find this amount...

Comment (mantikoros) [reply common/calculate-cpmm.ts:122]: at the very least, there definitely are ways to compute it more efficiently if this ever becomes the bottleneck (unlikely).

**Review by mantikoros — APPROVED** [review]
Very nice. A reasonably streamlined solution to adding limit orders based off the Bet object.

File: common/calculate-cpmm.ts (hunk 7 above)
Comment (jahooma) [reply common/calculate-cpmm.ts]: Ok, added a comment to that effect.

File: common/bet.ts (hunk 6 above)
Comment (jahooma) [reply common/bet.ts:44]: Oh, I guess I can fix that then. Maybe after my tests with discord people so as not to break things. Or, maybe I'll just leave it haha. Two L's makes more sense to me.

File: common/new-bet.ts (hunk 4 above)
Comment (jahooma) [reply common/new-bet.ts:178]: I guess I'll leave it for v2. I actually was planning to do it before.

**Pull request merged**